pennylane/transforms/gradients/qubit_parameter_shift.py

```diff
@@ -0,0 +1,151 @@
+            parameter :math:`\phi_k`, the nested list contains elements of the form
+            :math:`[c_i, a_i, s_i]` where :math:`i` is the index of the
+            term, resulting in a gradient recipe of
+
+            .. math:: \frac{\partial}{\partial\phi_k}f = \sum_{i} c_i f(a_i \phi_k + s_i).
+
+            If ``None``, the default gradient recipe containing the two terms
+            :math:`[c_0, a_0, s_0]=[1/2, 1, \pi/2]` and :math:`[c_1, a_1,
+            s_1]=[-1/2, 1, -\pi/2]` is assumed for every parameter.
+
+    Returns:
+        tuple[list[QuantumTape], function]: A tuple containing the list of generated tapes,
+        in addition to a post-processing function to be applied to the evaluated
+        tapes.
+
+    **Gradients of expectation values**
+
+    For a variational evolution :math:`U(\mathbf{p}) \vert 0\rangle` with
+    :math:`N` parameters :math:`\mathbf{p}`,
+
+    consider the expectation value of an observable :math:`O`:
+
+    .. math::
+
+        f(\mathbf{p})  = \langle \hat{O} \rangle(\mathbf{p}) = \langle 0 \vert
+        U(\mathbf{p})^\dagger \hat{O} U(\mathbf{p}) \vert 0\rangle.
+
+
+    The gradient of this expectation value can be calculated using :math:`2N` expectation
+    values using the parameter-shift rule:
+
+    .. math::
+
+        \frac{\partial f}{\partial \mathbf{p}} = \frac{1}{2\sin s} \left[ f(\mathbf{p} + s) -
+        f(\mathbf{p} -s) \right].
+
+    """
+    # check if the quantum tape contains any variance measurements
+    var_mask = [m.return_type is qml.operation.Variance for m in tape.measurements]
+
+    if any(var_mask):
+        raise ValueError("Does not support gradients of tapes with variance output.")
+
+    t_idx = list(tape.trainable_params)[idx]
+    op = tape._par_info[t_idx]["op"]
+    p_idx = tape._par_info[t_idx]["p_idx"]
+
+    if gradient_recipe is None:
+        gradient_recipe = op.get_parameter_shift(p_idx, shift=shift)
+
+    params = qml.math.stack(tape.get_parameters())
+    shift = np.zeros(qml.math.shape(params))
+    coeffs = []
+    tapes = []
+
+    for c, a, s in gradient_recipe:
+        shift[idx] = s
+        shifted_tape = tape.copy(copy_operations=True)
+        shifted_params = a * params + qml.math.convert_like(shift, params)
+        shifted_tape.set_parameters(qml.math.unstack(shifted_params))
+
+        coeffs.append(c)
+        tapes.append(shifted_tape)
+
+    def processing_fn(results):
+        """Computes the gradient of the parameter at index idx via the
+        parameter-shift method.
+
+        Args:
+            results (list[real]): evaluated quantum tapes
+
+        Returns:
+            array[float]: 1-dimensional array of length determined by the tape output
+            measurement statistics
+        """
+        results = qml.math.squeeze(qml.math.stack(results))
+        return sum([c * r for c, r in zip(coeffs, results)])
+
+    return tapes, processing_fn
+
+
+def hamiltonian_grad(tape, idx):

```

Comment:
@mariaschuld and here is the Hamiltonian gradient recipe. Here is my test script that works with this branch:

```python
import pennylane as qml
from pennylane import numpy as np
from pennylane.interfaces.batch import batch_execute
import tensorflow as tf


dev = qml.device("default.qubit", wires=3)

obs = [
    qml.PauliZ(0),
    qml.PauliX(0) @ qml.PauliZ(1),
    qml.PauliY(2)
]


def cost_fn(weights, x, fn="autograd"):
    H = qml.Hamiltonian(qml.math.stack([qml.math.sin(x), qml.math.cos(x), qml.math.tan(x)]), obs)

    with qml.tape.QuantumTape() as tape:
        qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1, 2])
        qml.expval(H)

    tapes, f = qml.transforms.hamiltonian_expand(tape)
    res = batch_execute(tapes,  dev, interface=fn)
    return f(res)

    return batch_execute([tape], dev, interface=fn)[0][0]


weights = np.ones([1, 3, 3], dtype=np.float32, requires_grad=True)
x = np.array(0.1)

res = cost_fn(weights, x, fn="autograd")
print("\n Result =", res.__repr__())

# weights.requires_grad = False
grad = qml.grad(cost_fn, argnum=0)(weights, x, fn="autograd")
print("\n Gradient (weights) =", grad.__repr__())

grad = qml.grad(cost_fn, argnum=1)(weights, x, fn="autograd")
print("\n Gradient (x) =", grad.__repr__())

hess = qml.jacobian(qml.grad(cost_fn, argnum=1), argnum=1)(weights, x, fn="autograd")
print("\n", hess.__repr__())



weights = tf.Variable(tf.ones([1, 3, 3], dtype=tf.float64))
x = tf.Variable(0.1, dtype=tf.float64)

with tf.GradientTape(persistent=True) as tape1:
    with tf.GradientTape() as tape2:
        before = dev.num_executions
        res = cost_fn(weights, 1.0 * x, fn="tf")
        after = dev.num_executions

    print("\n Result =", res, after-before)

    before = dev.num_executions
    grad = tape2.gradient(res, [weights, x])
    after = dev.num_executions
    print("\n Gradient (weights) =", grad, after-before)

before = dev.num_executions
hess = tape1.jacobian(grad[1], x, experimental_use_pfor=False)
after = dev.num_executions
print("\n", hess, after-before)
```